From 17aadc66bda1c0c85a739bc861280fd809d856f7 Mon Sep 17 00:00:00 2001
From: Tony Lin <tony.lin@freescale.com>
Date: Tue, 5 Jul 2011 14:36:14 +0800
Subject: [PATCH] ENGR00152560 [MX233/MX28]mxs-mmc: fix clock rate setting

Fix clock rate setting on mxs-mmc driver.
Previously, if div2 was zero the value for TIMING_CLOCK_RATE
would have been 255 instead of 0.
Also the limits for div1 (TIMING_CLOCK_DIVIDE) and
div2 (TIMING_CLOCK_RATE + 1) where not correctly defined.

Can easily be reproduced on mx23evk: default clock for high speed
sdio cards is 50 MHz.
With a SSP_CLK of 28.8 MHz (default), this resulted in an actual
clock rate of about 56 kHz.

Signed-off-by: Tony Lin <tony.lin@freescale.com>
---
 drivers/mmc/host/mxs-mmc.c |   14 +++++++-------
 1 files changed, 7 insertions(+), 7 deletions(-)

diff --git a/drivers/mmc/host/mxs-mmc.c b/drivers/mmc/host/mxs-mmc.c
index bd5e474..f94b717 100644
--- a/drivers/mmc/host/mxs-mmc.c
+++ b/drivers/mmc/host/mxs-mmc.c
@@ -3,7 +3,7 @@
  * Portions copyright (C) 2004-2005 Pierre Ossman, W83L51xD SD/MMC driver
  *
  * Copyright 2008 Embedded Alley Solutions, Inc.
- * Copyright 2009-2010 Freescale Semiconductor, Inc.
+ * Copyright 2009-2011 Freescale Semiconductor, Inc.
  *
  * This program is free software; you can redistribute it and/or modify
  * it under the terms of the GNU General Public License as published by
@@ -768,20 +768,20 @@ mxs_set_sclk_speed(struct mxs_mmc_host *host, unsigned int hz)
 	 */
 	ssp = clk_get_rate(host->clk);
 
-	for (div1 = 2; div1 < 254; div1 += 2) {
+	for (div1 = 2; div1 <= 254; div1 += 2) {
 		div2 = ssp / hz / div1;
-		if (div2 < 0x100)
+		if (div2 <= 256)
 			break;
 	}
-	if (div1 >= 254) {
+	if (div1 > 254) {
 		dev_err(host->dev, "Cannot set clock to %dHz\n", hz);
 		return;
 	}
 
 	if (div2 == 0)
-		bus_clk = ssp / div1;
-	else
-		bus_clk = ssp / div1 / div2;
+		div2 = 1;
+
+	bus_clk = ssp / div1 / div2;
 
 	dev_dbg(host->dev, "Setting clock rate to %ld Hz [%x+%x] "
 		"(requested %d), source %ldk\n",
-- 
1.5.4.4

